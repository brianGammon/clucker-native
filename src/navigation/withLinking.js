/* @flow */
import React, { Component } from 'react';
import { Linking } from 'react-native';
import { connect } from 'react-redux';
import { type NavigationContainer } from 'react-navigation';
import { removeInitialUrl as removeInitialUrlAction } from '../redux/actions';
import { type Navigation } from '../types';

const urlToPathAndParams = (uriPrefix, url) => {
  const params = {};
  const delimiter = uriPrefix || '://';
  let path = url.split(delimiter)[1];
  if (!path) {
    path = url;
  }
  return {
    path,
    params,
  };
};

type Props = {
  navigation: Navigation,
  initialUrl: string,
  removeInitialUrl: () => void,
};

export default (Comp: NavigationContainer<any, any, any>, uriPrefix: string) => {
  const { router } = Comp;

  class WrappedComponent extends Component<Props> {
    static router = router;

    componentDidMount() {
      // TODO: move to redux state instead of AsyncStorage
      const { initialUrl, removeInitialUrl } = this.props;
      if (initialUrl) {
        this.handleOpenURL({ url: initialUrl });
        removeInitialUrl();
      }
      Linking.addEventListener('url', this.handleOpenURL);
    }

    componentWillUnmount() {
      Linking.removeEventListener('url', this.handleOpenURL);
    }

    handleOpenURL = ({ url }) => {
      const parsedUrl = urlToPathAndParams(uriPrefix, url);
      const { navigation } = this.props;
      if (parsedUrl) {
        const { path, params } = parsedUrl;
        const action = router.getActionForPathAndParams(path, params);
        if (action) {
          if (action.action) {
            if (action.action.action) {
              navigation.dispatch(action.action.action);
            } else {
              navigation.dispatch(action.action);
            }
          } else {
            navigation.dispatch(action);
          }
        }
      }
    };

    render() {
      const { navigation } = this.props;
      return <Comp navigation={navigation} />;
    }
  }

  const mapStateToProps = ({ initialUrl }) => ({ initialUrl });
  const mapDispatchToProps = dispatch => ({
    removeInitialUrl: () => dispatch(removeInitialUrlAction()),
  });

  return connect(
    mapStateToProps,
    mapDispatchToProps,
  )(WrappedComponent);
};
