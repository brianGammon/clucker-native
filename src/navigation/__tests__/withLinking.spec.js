import React from "react";
import { Text, Linking } from "react-native";
import configureStore from "redux-mock-store";
import withLinking from "../withLinking";

const middlewares = [];
const mockStore = configureStore(middlewares);

const mockNavigation = {
  dispatch: jest.fn()
};

const action = {
  action1: "test",
  action: {
    action2: "test2",
    action: {
      action3: "test3"
    }
  }
};

const router = {
  getActionForPathAndParams: jest.fn()
};

beforeEach(() => {
  mockNavigation.dispatch.mockClear();
  router.getActionForPathAndParams = jest.fn();
});

test("withLinking wraps navigation component", () => {
  const store = mockStore({
    initialUrl: "clucker://test"
  });

  router.getActionForPathAndParams = jest.fn(() => action.action.action);
  const mock = () => <Text>Test Component</Text>;
  mock.router = router;

  // No uriPrefix for coverage
  const MockComponent = withLinking(mock);
  const wrapper = shallow(
    <MockComponent store={store} navigation={mockNavigation} />
  ).dive();
  expect(wrapper).toMatchSnapshot();
  expect(router.getActionForPathAndParams).toBeCalledWith("test", {});
  // Tests have a 1 level deep navigation action
  expect(mockNavigation.dispatch).toBeCalledWith(action.action.action);
  // For coverage of componentWillUnmount
  wrapper.instance().componentWillUnmount();
});

test("No initialUrl", () => {
  const store = mockStore({
    initialUrl: ""
  });

  const mock = () => <Text>Test Component</Text>;
  mock.router = router;

  const MockComponent = withLinking(mock, "clucker://");
  shallow(<MockComponent store={store} navigation={mockNavigation} />).dive();
  expect(Linking.addEventListener).toBeCalledWith("url", expect.anything());
  expect(router.getActionForPathAndParams).not.toBeCalled();
  expect(mockNavigation.dispatch).not.toBeCalled();
});

test("No prefix in initialUrl, and double action", () => {
  const store = mockStore({
    initialUrl: "test/test2"
  });

  router.getActionForPathAndParams = jest.fn(() => action.action);
  const mock = () => <Text>Test Component</Text>;
  mock.router = router;

  const MockComponent = withLinking(mock, "clucker://");
  shallow(<MockComponent store={store} navigation={mockNavigation} />).dive();
  expect(router.getActionForPathAndParams).toBeCalledWith("test/test2", {});
  expect(mockNavigation.dispatch).toBeCalledWith(action.action.action);
});

test("Triple action", () => {
  const store = mockStore({
    initialUrl: "test/test2/test3"
  });

  router.getActionForPathAndParams = jest.fn(() => action);

  const mock = () => <Text>Test Component</Text>;
  mock.router = router;

  const MockComponent = withLinking(mock, "clucker://");
  shallow(<MockComponent store={store} navigation={mockNavigation} />).dive();
  expect(mockNavigation.dispatch).toBeCalledWith(action.action.action);
});

test("No action found", () => {
  const store = mockStore({
    initialUrl: "someBadPath"
  });

  const mock = () => <Text>Test Component</Text>;
  mock.router = router;

  const MockComponent = withLinking(mock, "clucker://");
  shallow(<MockComponent store={store} navigation={mockNavigation} />).dive();
  expect(mockNavigation.dispatch).not.toBeCalled();
});
