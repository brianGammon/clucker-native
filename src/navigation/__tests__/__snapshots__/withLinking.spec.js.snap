// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withLinking wraps navigation component 1`] = `
<mock
  navigation={
    Object {
      "dispatch": [MockFunction] {
        "calls": Array [
          Array [
            Object {
              "action3": "test3",
            },
          ],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      },
    }
  }
/>
`;
